Reset replica engine before primary-replica resync

dnhatn opened the pull request:
When a replica starts following a newly promoted primary, it may have
some operations which don't exist on the new primary. We need to reset
replicas to the global checkpoint before executing primary-replica
resync. These two steps will align replicas to the primary.

This change resets an engine of a replica to the safe commit when
detecting a new primary term, then reindex operations from the local
translog up to the global checkpoint.

elasticmachine commented:
Pinging @elastic/es-distributed

dnhatn commented:
/cc @not-napoleon

s1monw commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -664,18 +672,17 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
-        return applyIndexOperation(seqNo, operationPrimaryTerm, version, null, autoGeneratedTimeStamp, isRetry,
-            Engine.Operation.Origin.REPLICA, sourceToParse);
+                                                           boolean isRetry, SourceToParse sourceToParse) throws IOException {
+        return applyIndexOperation(getEngine(), seqNo, operationPrimaryTerm, version, null, autoGeneratedTimeStamp,
```
I was wondering why you did that and I think I understand what you are trying to do. You try to make sure we always get the latest engine ie. the locked down one if we swap it. but there is still a race imo. inside applyIndexOperation you might have an engine that is already closed unless you put a lock around it. The swap might be atomic but the reference might still receive writes after you locked it down, is this ok?

dnhatn replied:
Before swapping engines, we drain all IndexShardOperationPermits (backed by Semaphore) and a write operation requires an IndexShardPermit. I think we are okay here.

ywelsch replied:
correct.

s1monw replied:
++ maybe we add this as a comment somewhere or even as an assertion?

ywelsch replied:
Don't we need to use `getEngineForResync` here? Assume that there are documents already replicated by the new primary before this replica has received all resync operations. Also wondering why there have been no test failures, maybe test coverage is not good? Also note that `getEngineForResync` is probably not the best name for this. I think there's a bigger issue here, let's sync about this tomorrow.

s1monw commented on server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -2234,6 +2240,22 @@ public SeqNoStats getSeqNoStats(long globalCheckpoint) {
     }
 
+    @Override
+    public Engine lockDownEngine() throws IOException {
```
ideally we would do this entirely outside of the engine and maybe just pass and engine to the ctor of ReadOnlyEngine? do we need to make sure we don't receive writes after we did this or why do we acquire a write lock?

dnhatn replied:
I will move this to ctor of ReadOnlyEngine.

dnhatn commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -196,6 +196,11 @@
     protected volatile long operationPrimaryTerm;
     protected final AtomicReference<Engine> currentEngineReference = new AtomicReference<>();
+
+    private final AtomicReference<Engine> resettingEngineReference = new AtomicReference<>();
```
+1

ywelsch commented on server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java:
```diff
@@ -0,0 +1,314 @@
+
+    @Override
+    public void refresh(String source) throws EngineException {
+        // noop
```
refreshes should not be happening? If so, should we throw an UnsupportedOperationException here?

ywelsch commented on server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java:
```diff
@@ -0,0 +1,314 @@
+
+    @Override
+    public CommitId flush(boolean force, boolean waitIfOngoing) throws EngineException {
+        throw new UnsupportedOperationException();
```
do we want to assert that all of these methods are never called?

ywelsch commented on server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java:
```diff
@@ -0,0 +1,314 @@
+/**
+ * An engine that does not accept writes, and always points stats, searcher to the last commit.
+ */
+final class ReadOnlyEngine extends Engine {
```
I wonder if we should give this a different name, in particular because we might have something similar for frozen indices. There it might be a more complete version of `readonly`, with possibility to take a `Translog.Snapshot`. Maybe we could call this `SearchOnlyEngine`.

dnhatn commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2630,4 +2660,90 @@ public void afterRefresh(boolean didRefresh) throws IOException {
+
+    final boolean canResetEngine() {
+        // TODO: do not reset the following shard
+        return indexSettings.getIndexVersionCreated().onOrAfter(Version.V_6_4_0);
```
@ywelsch Yeah, you understood it correctly. I had the same thought but did not go with that option as I wasn't sure if it's a right trade-off. I am glad that you suggest it. Should we make that change into this PR or a separate prerequisite PR to reduce noise in this PR?

dnhatn commented:
@s1monw and @ywelsch It's ready for another round. Can you please take a look? Thank you!

s1monw commented on server/src/main/java/org/elasticsearch/index/engine/SearchOnlyEngine.java:
```diff
@@ -0,0 +1,315 @@
+
+    @Override
+    protected void closeNoLock(String reason, CountDownLatch closedLatch) {
+        try {
```
I think you should protect this against double counting down the `closeLatch` by wrapping this entire try block in 

```Java
if (isClosed.compareAndSet(false, true)) {

}
```

s1monw commented on server/src/main/java/org/elasticsearch/index/engine/SearchOnlyEngine.java:
```diff
@@ -0,0 +1,315 @@
+        this.seqNoStats = engine.getSeqNoStats(engine.getLastSyncedGlobalCheckpoint());
+        this.translogStats = engine.getTranslogStats();
+        this.lastCommittedSegmentInfos = engine.getLastCommittedSegmentInfos();
+        Searcher searcher = engine.acquireSearcher("lockdown", SearcherScope.INTERNAL);
```
can you leave a comment here that we keep a reference to the store implicitly through the searcher? I do wonder if we should make it explicit

s1monw commented on server/src/main/java/org/elasticsearch/index/engine/SearchOnlyEngine.java:
```diff
@@ -0,0 +1,315 @@
+        this.lastCommittedSegmentInfos = engine.getLastCommittedSegmentInfos();
+        Searcher searcher = engine.acquireSearcher("lockdown", SearcherScope.INTERNAL);
+        try {
+            this.searcherManager = new SearcherManager(searcher.getDirectoryReader(),
```
this searcher manager seems to be unclosed. I think you should close it as well in the closeNoLock method?

s1monw commented on server/src/main/java/org/elasticsearch/index/engine/SearchOnlyEngine.java:
```diff
@@ -0,0 +1,315 @@
+        store.incRef();
+        Releasable releasable = store::decRef;
+        try (ReleasableLock ignored = readLock.acquire()) {
+            final EngineSearcher searcher = new EngineSearcher(source, searcherManager, store, logger);
```
can you try to exercise this method to make sure we open a new searcher and close / release everything

dnhatn replied:
Ah, `getDocIds` method in `SearchOnlyEngineTests#testSearchOnlyEngine` acquires searchers.

s1monw commented on server/src/main/java/org/elasticsearch/index/engine/SearchOnlyEngine.java:
```diff
@@ -0,0 +1,315 @@
+    private final SearcherManager searcherManager;
+    private final Searcher lastCommitSearcher;
+
+    public SearchOnlyEngine(Engine engine) throws IOException {
```
I do wonder if it would make more sense to open this entire thing off a store directly and maybe just pass and EngineConfig to this. it would make it more generic and less bound to an engine. WDYT?

s1monw replied:
something like this:

```Java

 public SearchOnlyEngine(EngineConfig config) {
        super(config);
        try {
            Store store = config.getStore();
            store.incRef();
            DirectoryReader reader = null;
            boolean success = false;
            try {
                this.lastCommittedSegmentInfos = Lucene.readSegmentInfos(store.directory());
                this.translogStats = new TranslogStats(0, 0, 0, 0, 0);
                final SequenceNumbers.CommitInfo seqNoStats =
                    SequenceNumbers.loadSeqNoInfoFromLuceneCommit(lastCommittedSegmentInfos.userData.entrySet());
                long maxSeqNo = seqNoStats.maxSeqNo;
                long localCheckpoint = seqNoStats.localCheckpoint;
                this.seqNoStats = new SeqNoStats(maxSeqNo, localCheckpoint, localCheckpoint);
                reader = SeqIdGeneratingDirectoryReader.wrap(ElasticsearchDirectoryReader.wrap(DirectoryReader
                .open(store.directory()), config.getShardId()), config.getPrimaryTermSupplier().getAsLong());
                this.indexCommit = reader.getIndexCommit();
                this.searcherManager = new SearcherManager(reader, new SearcherFactory());
                success = true;
            } finally {
                if (success == false) {
                    IOUtils.close(reader, store::decRef);
                }
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e); // this is stupid
        }
    }
```

I did something similar a while back so I had it ready... I am not sure it safe to use 💯

dnhatn replied:
I adopted this, but I have to pass SeqNoStats from outside because we use a "reset" local checkpoint which may not equal to the value from an index commit.

s1monw commented on server/src/main/java/org/elasticsearch/index/engine/SearchOnlyEngine.java:
```diff
@@ -0,0 +1,315 @@
+
+    public SearchOnlyEngine(Engine engine) throws IOException {
+        super(engine.config());
+        engine.refresh("lockdown");
```
we should do this refresh outside no? and it should be a flush?

s1monw commented on server/src/main/java/org/elasticsearch/index/engine/SearchOnlyEngine.java:
```diff
@@ -0,0 +1,315 @@
+
+    @Override
+    public void maybePruneDeletes() {
+
```
nit extra newline

s1monw commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1266,14 +1269,16 @@ public void trimOperationOfPreviousPrimaryTerms(long aboveSeqNo) {
     int runTranslogRecovery(Engine engine, Translog.Snapshot snapshot) throws IOException {
-        recoveryState.getTranslog().totalOperations(snapshot.totalOperations());
-        recoveryState.getTranslog().totalOperationsOnStart(snapshot.totalOperations());
+        if (isEngineResetting() == false) {
```
any reason we can't just run this the same way we do if we are not resetting?

s1monw commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2140,33 +2142,22 @@ public void onFailedEngine(String reason, @Nullable Exception failure) {
+    private Engine createNewEngine(EngineConfig config) throws IOException {
+        assert Thread.holdsLock(mutex);
+        if (state == IndexShardState.CLOSED) {
+            throw new AlreadyClosedException(shardId + " can't create engine - shard is closed");
```
not sure, should we through IndexShardNotClosedException instead?

dnhatn replied:
Yes, we should throw IndexShardNotClosedException. AlreadyClosedException was a left-over when we folded Engine to IndexShard.

s1monw commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2648,4 +2645,148 @@ public void afterRefresh(boolean didRefresh) throws IOException {
+        }
+
+        @Override
+        public void close() throws IOException {
```
make this synchronized too. it's safer since you modify both references

dnhatn replied:
good catch!

s1monw commented on server/src/main/java/org/elasticsearch/index/translog/Translog.java:
```diff
@@ -555,21 +555,49 @@ public long getLastSyncedGlobalCheckpoint() {
+                .filter(reader -> reader.getGeneration() >= minTranslogFileGen && reader.getCheckpoint().minSeqNo <= upToSeqNo)
                 .map(BaseTranslogReader::newSnapshot).toArray(TranslogSnapshot[]::new);
-            return newMultiSnapshot(snapshots);
+            Snapshot snapshot = newMultiSnapshot(snapshots);
```
maybe you just return snapshot if upToSeqNo is == Long.MAX_VALUE?

dnhatn replied:
Done

dnhatn commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2662,85 +2629,147 @@ public void afterRefresh(boolean didRefresh) throws IOException {
+        }
+
+        @Override
+        public void close() throws IOException {
```
+1

dnhatn commented:
@s1monw and @ywelsch I've addressed your comments. Can you please give it another go? I will beef up integration tests as Yannick suggested.

ywelsch commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2648,4 +2641,161 @@ public void afterRefresh(boolean didRefresh) throws IOException {
+                return null;
+            } else {
+                engineHolder.makeActiveEngineSearchOnly();
+                final Engine pendingEngine = createNewEngine(newEngineConfig());
```
this trims unsafe commits, possibly cleaning up segments that are referenced by the active search only engine?

dnhatn replied:
We open a directory reader in the constructor of a search-only engine and keeps that reader until we manually close the search-only engine. Holding that reader would prevent the segment files of the last commit from deleting during trimming unsafe commits.

ywelsch commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2648,4 +2641,161 @@ public void afterRefresh(boolean didRefresh) throws IOException {
+
+    private void completePendingEngineWithLocalHistory() throws IOException {
+        final Engine pendingEngine;
+        synchronized (mutex) {
```
why do you need to do this under the mutex?

ywelsch commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2648,4 +2641,161 @@ public void afterRefresh(boolean didRefresh) throws IOException {
+                    this.pendingEngine = null;
+                }
+            }
+            IOUtils.close(closing);
```
How do we ensure that searches are not accessing acquireSearcher on the closed engine and switching to the new engine? Also, is there a test that checks that searches (with preference set to this node) continue to work during this transition.

dnhatn replied:
There is a small interval in that callers might acquire searchers from the closed engine and hit AlreadClosedException. We can void this entirely by retrying on "AlreadClosedException" if the accessing engine is different from the current engine. However, I am not sure if we should do it.

There is a test which continuously acquires searchers and makes sure that all acknowledged writes are maintained during the transition. https://github.com/elastic/elasticsearch/pull/32867/files#diff-b268f5fefa5837ece96b957e46f628cbR674 (getShardDocUIDs acquires a searcher and uses that searcher to collect document Ids).

ywelsch replied:
> However, I am not sure if we should do it.

why is that? We're building all this machinery to have search availability during the transition, except for this very short moment?

I had the same idea about retrying. An alternative would be to do refcounting for closing the engine, to ensure that we only actually close once all in-flight `acquireSearcher` calls have been completed.

dnhatn replied:
Sorry, I did not think this carefully. I thought we have to implement the retry on all methods that we support on the search-only, but I was wrong. We only need to implement retry for "get" and "acquire searcher". These two methods should be simple. Thanks for this great question.

dnhatn commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2296,13 +2283,19 @@ public void acquireReplicaOperationPermit(final long opPrimaryTerm, final long g
-                                getEngine().resetLocalCheckpoint(localCheckpoint);
-                                getEngine().rollTranslogGeneration();
+                                if (isEngineResetting()) {
+                                    engineHolder.closePendingEngine();
```
I think this is okay because we start another engine after that in this case. Moreover, we stick the "reset" local checkpoint (expose the local checkpoint of the active engine) while resetting the engine; thus the global checkpoint won't advance.

dnhatn commented:
@ywelsch I fixed the mutex and logging issue, and added a new assertion to `AbstractDisruptionTestCase` and `ClusterDisruptionIT#testAckedIndexing`. However, I am still working on an IntegTest.

ywelsch commented on server/src/test/java/org/elasticsearch/discovery/AbstractDisruptionTestCase.java:
```diff
@@ -93,6 +94,11 @@ public void setUp() throws Exception {
 
+    @After
+    public void assertSameDocIds() throws Exception {
+        assertSameDocIdsOnShards();
```
There's already a method called `assertSeqNos`. Should we call that one first? can you also check `assertSameDocIdsOnShards` for RelocationIT?

ywelsch commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -969,20 +974,60 @@ public void onFailure(final Exception e) {
     }
 
+    public void testDoNoResetEngineOnOldVersions() throws Exception {
+        Version oldVersion = randomValueOtherThanMany(v -> v.onOrAfter(Version.V_6_4_0),
```
VersionUtils has a method for this

ywelsch commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -2129,6 +2132,30 @@ protected NumShards getNumShards(String index) {
+    /**
+     * Asserts that all shards with the same shardId should have document Ids.
+     */
+    public void assertSameDocIdsOnShards() throws Exception {
```
do we need to wrap this in an assertBusy?

dnhatn replied:
Done

ywelsch commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -2129,6 +2132,30 @@ protected NumShards getNumShards(String index) {
+                    try {
+                        final Set<String> docIds = IndexShardTestCase.getShardDocUIDs(shard, true);
+                        if (docIdsPerShard.containsKey(shard.shardId())) {
+                            assertThat(docIds, equalTo(docIdsPerShard.get(shard.shardId())));
```
If this fails, we will need more output. Can you add information about which two shard copies diverged?

dnhatn commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1343,13 +1346,9 @@ private void innerOpenEngineAndTranslog() throws IOException {
-
-        createNewEngine(config);
+        synchronized (mutex) {
+            engineHolder.activateNewEngine(createNewEngine(config));
```
We need to create a new engine under mutex, otherwise, we might clean up an index commit which is being snapshotted by `IndexShard#snapshotStoreMetadata` (without an engine).

dnhatn commented:
This approach might prevent acknowledged writes after a promotion but before a resync from being exposed. We discussed this and decided to go with a more contained solution (suggested by Yannick) even though some acknowledged writes may disappear during the primary-replica resync with the new approach.

@ywelsch and @s1monw It is ready for another round. It's easier to grab than the previous ones. Thanks so much for all iterations on this.

s1monw commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2648,4 +2662,34 @@ public void afterRefresh(boolean didRefresh) throws IOException {
+            logger.info("resetting replica engine from max_seq_no [{}] to seq_no [{}] global checkpoint [{}]",
+                seqNoStats.getMaxSeqNo(), upToSeqNo, seqNoStats.getGlobalCheckpoint());
+            final TranslogStats translogStats = currentEngine.getTranslogStats();
+            final SearchOnlyEngine searchOnlyEngine = new SearchOnlyEngine(currentEngine.config(), seqNoStats, translogStats);
```
I think it's ok to do this under the mutex since it's quick but why do we need to do this?

s1monw commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2296,13 +2307,16 @@ public void acquireReplicaOperationPermit(final long opPrimaryTerm, final long g
+                                logger.info("detected new primary with primary term [{}], resetting local checkpoint from [{}] to [{}]",
+                                    opPrimaryTerm, getLocalCheckpoint(), localCheckpoint);
+                                if (localCheckpoint < getEngine().getSeqNoStats(localCheckpoint).getMaxSeqNo()) {
+                                    getEngine().flush(true, true);
```
can we get the engine only once in both cases ?

s1monw commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1447,7 +1469,8 @@ private void ensureWriteAllowed(Engine.Operation.Origin origin) throws IllegalIn
 
         if (origin.isRecovery()) {
-            if (state != IndexShardState.RECOVERING) {
+            final boolean isResetting = getEngineOrNull() instanceof SearchOnlyEngine;
```
I am not a super fan of this. I wonder if we can afterwards rethink this.

ywelsch replied:
same here.
I would prefer to have a separate `runTranslogRecovery` method that we use for the local replay after reset. This method could then use a different stats object as well as a different origin.

s1monw commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1331,30 +1361,22 @@ private void innerOpenEngineAndTranslog() throws IOException {
+        final EngineConfig config = newEngineConfig();
+        // we disable deletes since we allow for operations to be executed against the shard while recovering
+        // but we need to make sure we don't loose deletes until we are done recovering
+        config.setEnableGcDeletes(false);
```
do we need to call `  engine.onSettingsChanged();` after this?

s1monw commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1266,14 +1296,14 @@ public void trimOperationOfPreviousPrimaryTerms(long aboveSeqNo) {
     // package-private for testing
     int runTranslogRecovery(Engine engine, Translog.Snapshot snapshot) throws IOException {
-        recoveryState.getTranslog().totalOperations(snapshot.totalOperations());
+        recoveryState.getTranslog().setOrIncreaseTotalOperations(snapshot.totalOperations());
```
I think we need to have dedicated stats. I assume we should do it based on if we are resetting or not. We can do this in a followup.

s1monw commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -500,15 +503,21 @@ public void updateShardState(final ShardRouting newRouting,
                                  */
-                                engine.rollTranslogGeneration();
-                                engine.fillSeqNoGaps(newPrimaryTerm);
+                                getEngine().rollTranslogGeneration();
```
can we fetch the engine only once in this method?

ywelsch commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -500,15 +503,21 @@ public void updateShardState(final ShardRouting newRouting,
                                  */
-                                final Engine engine = getEngine();
-                                engine.restoreLocalCheckpointFromTranslog();
+                                if (seqNoStats().getLocalCheckpoint() < maxSeqNoOfResettingEngine) {
```
having this `maxSeqNoOfResettingEngine` variable is messy. Can't we get the max sequence number information from the translog? We essentially want to check if Lucene needs to catch up with what's in the translog.

ywelsch commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -500,15 +503,21 @@ public void updateShardState(final ShardRouting newRouting,
+                                    // The engine was reset before but hasn't restored all existing operations yet.
+                                    // We need to reset the engine again with all the local history.
+                                    getEngine().flush();
+                                    resetEngineUpToSeqNo(Long.MAX_VALUE);
```
I don't think we should use `resetEngineUpToSeqNo` here with all its machinery of `SearchOnlyEngine` etc. which only makes sense if there are conflicts in the history. As a primary, we dictate the history here, and can therefore "just" recover from translog (+ refresh afterwards)?

ywelsch commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1184,6 +1190,26 @@ private void markSearcherAccessed() {
         }
     }
 
+    private <T> T accessEngineAndRetryOnSwap(Function<Engine, T> accessor) {
```
this should indeed be extremely unlikely to cause a user-visible issue (as at least the new primary should not run into this). Still, it's not great if an active shard cannot serve searches at all times. Let's take this out of this PR for now and rethink this later.

dnhatn commented:
Discussed this with @bleskes  on another channel. We are going to split this PR into 3 smaller PRs so we can review.

- [ ] Reset replica engine to the global checkpoint on promotion
- [ ] Restore local history on the new primary upon promotion. This is to deal with cascading fail-over.
- [ ] Support search on replicas during promotion

dnhatn commented:
All subtasks of this PR were done and merged. I am closing this.
